module/synchronization/core.go

```diff
@@ -47,31 +48,52 @@ func DefaultConfig() Config {
 //
 // Core is safe for concurrent use by multiple goroutines.
 type Core struct {
-	log      zerolog.Logger
-	Config   Config
-	mu       sync.Mutex
-	heights  map[uint64]*Status
-	blockIDs map[flow.Identifier]*Status
+	log               zerolog.Logger
+	Config            Config
+	mu                sync.Mutex
+	heights           map[uint64]*Status
+	maxHeightReceived uint64
+	blockIDs          map[flow.Identifier]*statusWithHeight
+}
+
+var _ module.SyncCore = (*Core)(nil)
+
+type statusWithHeight struct {
+	*Status
+	height uint64
 }
 
 func New(log zerolog.Logger, config Config) (*Core, error) {
 	core := &Core{
-		log:      log.With().Str("module", "synchronization").Logger(),
-		Config:   config,
-		heights:  make(map[uint64]*Status),
-		blockIDs: make(map[flow.Identifier]*Status),
+		log:               log.With().Str("module", "synchronization").Logger(),
+		Config:            config,
+		heights:           make(map[uint64]*Status),
+		blockIDs:          make(map[flow.Identifier]*statusWithHeight),
+		maxHeightReceived: 0,
 	}
 	return core, nil
 }
 
+func (c *Core) setBlockIDStatus(blockID flow.Identifier, height uint64, status *Status) {
+	c.blockIDs[blockID] = &statusWithHeight{status, height}
+}
+
+func (c *Core) getBlockIDStatus(blockID flow.Identifier) *Status {
+	s, ok := c.blockIDs[blockID]
+	if !ok {
+		return nil
+	}
+	return s.Status
+}
+
 // HandleBlock handles receiving a new block from another node. It returns
 // true if the block should be processed by the compliance layer and false
 // if it should be ignored.
 func (c *Core) HandleBlock(header *flow.Header) bool {
```

Comment:
We should consider adjusting the naming or documentation here. This generically-named function now specifically assumes that the block was requested by ID (ie. if I request a block by height, then call `HandleBlock`, it will not be processed properly).

Similarly, `HandleFinalizedBlock` assumes the block was requested by height.